node/lib/util/push.js

```diff
@@ -84,23 +84,75 @@ exports.push = co.wrap(function *(repo, remoteName, source, target, force) {
     let remoteUrl = yield GitUtil.getUrlFromRemoteName(repo, remoteName);
 
     // First, push the submodules.
+    const annotatedCommit = yield GitUtil.resolveCommitish(repo, source);
```

Comment:
I'd like to see the logic for listing what to push factored out into a separate (and separately-tested) method.